internals/overlord/logstate/loki/loki.go

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package loki
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/internals/overlord/logstate/clienterr"
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+const maxRequestEntries = 1000
+
+var requestTimeout = 10 * time.Second
+
+type Client struct {
+	remoteURL string
+	// buffered entries are "sharded" by service name
+	entries map[string][]lokiEntry
+	// Keep track of the number of buffered entries, to avoid having to iterate
+	// the entries map to get the total number.
+	numEntries int
+
+	httpClient *http.Client
+	retryAfter *time.Time
+}
+
+func NewClient(target *plan.LogTarget) *Client {
+	return &Client{
+		remoteURL:  target.Location,
+		entries:    map[string][]lokiEntry{},
+		httpClient: &http.Client{Timeout: requestTimeout},
+	}
+}
+
+func (c *Client) Write(ctx context.Context, entry servicelog.Entry) error {
+	c.entries[entry.Service] = append(c.entries[entry.Service], encodeEntry(entry))
+	c.numEntries++
+
+	if c.numEntries >= maxRequestEntries {
+		if c.retryAfter != nil {
+			if time.Now().Before(*c.retryAfter) {
+				// Retry-After deadline hasn't passed yet, so we shouldn't flush
+				return nil
+			}
+			c.retryAfter = nil
+		}
+		return c.Flush(ctx)
+	}
+	return nil
+}
+
+func encodeEntry(entry servicelog.Entry) lokiEntry {
+	return lokiEntry{
+		strconv.FormatInt(entry.Time.UnixNano(), 10),
+		strings.TrimSuffix(entry.Message, "\n"),
+	}
+}
+
+func (c *Client) Flush(ctx context.Context) error {
```

Comment:
Just a small detail that isn't going to matter a great deal but should we also set the HTTP header here for the client? Just means anyone inspecting the protocol as it goes through all the places has some more context.

Reply:
What do you mean by "set the HTTP header here for the client"?

Reply:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent

Reply:
It's probably also not a bad idea to set the `Accept` header as well if we want to get anal about having good standards.

Reply:
totally happy to set the `User-Agent` and `Accept` headers here, but I'm not an expert on those, so I have no idea what a reasonable value is. What do you suggest?

Reply:
Normally you'd set the User-Agent to something like `pebble/{version}` (if the version is easily accessible). Not sure about Accept -- I wouldn't have thought it's necessary here for a known API request.

Reply:
Should the `os.Hostname()` go in `User-Agent` though?

Reply:
Nah, definitely not. It's the name/version of the client, not the host/IP.